src/hotspot/os/linux/os_linux.cpp

```diff
@@ -4832,6 +4843,16 @@ jint os::init_2(void) {
     }
   }
 
+  // Check the availability of MADV_POPULATE_WRITE.
+  if (FLAG_IS_DEFAULT(UseMadvPopulateWrite) && UseMadvPopulateWrite) {
+    // Some uek releases recognize MADV_POPULATE_WRITE_value as another advice,
+    // so a trick for uek releases is required.
+    // See https://github.com/oracle/linux-uek/issues/23
+    const int flag = MADV_POPULATE_WRITE + (os::Linux::_is_uek_release ? 1 : 0);
+    FLAG_SET_DEFAULT(UseMadvPopulateWrite, (::madvise(0, 0, flag) == 0));
```

Comment:
I read the supporting material and I still can't get this sorted in my head. From my reading some UEK versions have this problem and some do not. But you do this "trick" for all UEK versions and I just don't understand how that works. ??

Reply:
Personally, I think checking 24 for all UEK versions is safe:
- for versions prior to 5.15 UEK, 23 is supported as MADV_DONTEXEC, while 24 is not supported;
- for 5.15 (, 5.16 and 5.17 if existed) UEK, 23 is supported as MADV_POPULATE_WRITE, while 24 is supported as MADV_DOEXEC;
- for 5.18 or newer (with or without UEK), both 23 and 24 are supported.

Reply:
So what is the affect of using MADV_DOEXEC when the intent was to use MADV_POPULATE_WRITE?

Reply:
MADV_DOEXEC is not used in os::pd_pretouch_memory, but just used in os::init_2 for 5.15 UEK. The key is MADV_POPULATE_WRITE is supported as 23 on UEK releases if and only if 24 is supported. If the code is confusing, I will check whether the version of kernel is newer than 5.14 here rather than checking the support of 24.